multi: Remove need for DerivePrivKey from watchtower, brontide, netann and sphinx package

guggero opened the pull request:
This PR is a first step towards #3929 and satisfies the first two steps described there.

It was quite challenging to split this into smaller commits that still compile.
So please don't suggest changing the commit structure lightly ;-)

The PR is structured in 4 parts:
1. Prepare, add new interfaces.
2. Pure renaming to keep later changes in the logic as small and easy to digest as possible.
3. Switch to new interfaces by removing usage of private keys directly.
4. Clean up, remove now unused code.

Roasbeef commented:
Can be updated now that https://github.com/lightningnetwork/lightning-onion/pull/47 has been merged.

Roasbeef commented on keychain/derivation.go:
```diff
@@ -176,6 +176,10 @@ type KeyRing interface {
 type SecretKeyRing interface {
 	KeyRing
 
+	ECDHRing
```
Hmm, may be a bit too early to call, but not sure we want to _embed_ these interfaces 🤔

guggero replied:
Currently this is easiest to do, as we just pass in a `SecretKeyRing` in most places or use the `keyRing` of the `activeChainControl`. But we can decouple this further in a follow-up PR where we replace the key ring with something that's behind an RPC.

Roasbeef commented on keychain/ecdh.go:
```diff
@@ -0,0 +1,56 @@
+	"github.com/btcsuite/btcd/btcec"
+)
+
+func NewPubKeyECDH(keyDesc KeyDescriptor, ecdh ECDHRing) *PubKeyECDH {
```
Missing `godoc` comments.

Roasbeef commented on go.mod:
```diff
@@ -68,4 +68,6 @@ replace git.schwanenlied.me/yawning/bsaes.git => github.com/Yawning/bsaes v0.0.0
 // btcsuite/btcutil package requests a newer version.
 replace golang.org/x/crypto => golang.org/x/crypto v0.0.0-20190211182817-74369b46fc67
 
+replace github.com/lightningnetwork/lightning-onion => github.com/guggero/lightning-onion v1.0.2-0.20200428130542-d019eaf53e32
```
Can be updated to just point to onion master. We should also start to tag versions for the lib.

Roasbeef commented on brontide/noise.go:
```diff
@@ -289,8 +285,8 @@ type handshakeState struct {
-	localStatic    *btcec.PrivateKey
-	localEphemeral *btcec.PrivateKey
+	localStatic    keychain.SingleKeyECDH
+	localEphemeral keychain.SingleKeyECDH // nolint (false positive)
```
Just curious, what was the false lint warning?

guggero replied:
The linter thinks it's unused because `handshakeState` is embedded into `Machine` and the `localEphemeral` is only ever accessed through the `Machine` struct.
There is an issue still open in the `golangci-lint` repo about false positives for embedded structs: https://github.com/golangci/golangci-lint/issues/826

guggero commented:
Rebased and updated to include the merged lightningnetwork/lightning-onion#47.
I don't have any rights to push a tag to the `lightning-onion` repo so I added a commit dependency as usual.

Roasbeef commented on go.sum:
```diff
@@ -156,6 +160,8 @@ github.com/lightninglabs/protobuf-hex-display v1.3.3-0.20191212020323-b444784ce7
 github.com/lightninglabs/protobuf-hex-display v1.3.3-0.20191212020323-b444784ce75d/go.mod h1:KDb67YMzoh4eudnzClmvs2FbiLG9vxISmLApUkCa4uI=
 github.com/lightningnetwork/lightning-onion v1.0.1 h1:qChGgS5+aPxFeR6JiUsGvanei1bn6WJpYbvosw/1604=
 github.com/lightningnetwork/lightning-onion v1.0.1/go.mod h1:rigfi6Af/KqsF7Za0hOgcyq2PNH4AN70AaMRxcJkff4=
+github.com/lightningnetwork/lightning-onion v1.0.2-0.20200501022730-3c8c8d0b89ea h1:oCj48NQ8u7Vz+MmzHqt0db6mxcFZo3Ho7M5gCJauY/k=
```
Does a `go mod tidy` get rid of the extra lines here for the prior onion version?

guggero replied:
Ah yes, good catch! Fixed.

Roasbeef commented:
Needs another rebase.

guggero commented:
Rebased and linter silenced.

cfromknecht commented on watchtower/wtmock/keyring.go:
```diff
@@ -20,29 +20,6 @@ func NewSecretKeyRing() *SecretKeyRing {
-// DerivePrivKey derives the private key for a given key descriptor. If
-// this method is called twice with the same argument, it will return the same
-// private key.
-func (m *SecretKeyRing) DerivePrivKey(
```
👍
